tools/crm_mon.c

```diff
@@ -3652,47 +4156,111 @@ mon_refresh_display(gpointer user_data)
                 fprintf(stderr, "Critical: Unable to output html file\n");
                 clean_up(CRM_EX_CANTCREAT);
             }
             break;
 
         case mon_output_xml:
-            print_xml_status(&data_set);
+            print_xml_status(&data_set, stonith_history);
             break;
 
         case mon_output_monitor:
-            print_simple_status(&data_set);
+            print_simple_status(&data_set, stonith_history);
             if (has_warnings) {
                 clean_up(MON_STATUS_WARN);
             }
             break;
 
         case mon_output_plain:
         case mon_output_console:
-            print_status(&data_set);
+            print_status(&data_set, stonith_history);
             break;
 
         case mon_output_none:
             break;
     }
 
+    stonith_history_free(stonith_history);
+    stonith_history = NULL;
     cleanup_alloc_calculations(&data_set);
+
+    /* poll for pending fence-actions as we don't get notifications so far */
+    if ((!one_shot) && (show & mon_show_fence_history)) {
+        kick_refresh(FALSE);
+    }
     return TRUE;
 }
 
 void
-mon_st_callback(stonith_t * st, stonith_event_t * e)
+mon_st_callback_event(stonith_t * st, stonith_event_t * e)
 {
-    char *desc = crm_strdup_printf("Operation %s requested by %s for peer %s: %s (ref=%s)",
-                                 e->operation, e->origin, e->target, pcmk_strerror(e->result),
-                                 e->id);
+    if (st->state == stonith_disconnected) {
+        st->state = stonith_disconnected;
+        /* disconnect cib as well and have everything reconnect */
+        mon_cib_connection_destroy(NULL);
+    } else {
+        char *desc = crm_strdup_printf("Operation %s requested by %s for peer %s: %s (ref=%s)",
+                                    e->operation, e->origin, e->target, pcmk_strerror(e->result),
+                                    e->id);
 
-    if (external_agent) {
-        send_custom_trap(e->target, NULL, e->operation, pcmk_ok, e->result, 0, desc);
+        if (external_agent) {
+            send_custom_trap(e->target, NULL, e->operation, pcmk_ok, e->result, 0, desc);
+        }
+        free(desc);
+    }
+}
+
+void kick_refresh(gboolean data_updated)
+{
+    static int updates = 0;
+    long now = time(NULL);
+
+    if (data_updated) {
+        updates++;
+    }
+
+    if(refresh_timer == NULL) {
+        refresh_timer = mainloop_timer_add("refresh", 2000, FALSE, mon_trigger_refresh, NULL);
+    }
+
+    /* Refresh
+     * - immediately if the last update was more than 5s ago
+     * - every 10 cib-updates
+     * - at most 2s after the last update
+     */
+    if ((now - last_refresh) > (reconnect_msec / 1000)) {
+        mainloop_set_trigger(refresh_trigger);
+        mainloop_timer_stop(refresh_timer);
+        updates = 0;
+
+    } else if(updates > 10) {
+        mainloop_set_trigger(refresh_trigger);
+        mainloop_timer_stop(refresh_timer);
+        updates = 0;
```

Comment:
Technically this is every 11 updates :)

Reply:
True. Both the comment and the implementation are just moved here but probably code & comment should be synced.